### ISSUE TITLE
Required ruby version is incorrect

### ISSUE THREAD

**Issue by dduugg** [sorbet/sorbet#4265]
The [required ruby version](https://github.com/sorbet/sorbet/blob/master/gems/sorbet-runtime/sorbet-runtime.gemspec#L14) of `sorbet-runtime` is listed as `>= 2.3.0`.

However, the gem uses [endless ranges](https://github.com/sorbet/sorbet/blob/master/gems/sorbet-runtime/lib/types/utils.rb#L87), which is a Ruby 2.6 [feature](https://bugs.ruby-lang.org/issues/12912), breaking on earlier ruby releases.

Could I also humbly request that the test suite be updated to run on the latest patch of each supported minor ruby release (e.g. 2.6, 2.7)? Thanks!

**Comment by jez** [comment]
We document what versions of Ruby Sorbet supports here:

https://sorbet.org/docs/faq#what-platforms-does-sorbet-support

If you feel particularly strongly that we should capture that in the gemspec I can, but I tend to think that version ranges are brittle, because the true test is whether it works, not whether the versions match.

> Could I also humbly request that the test suite be updated to run on the latest patch of each supported minor ruby release (e.g. 2.6, 2.7)? Thanks!

The sorbet-runtime test suite already runs against both ruby 2.6 and ruby 2.7.

**Comment by dduugg** [comment]
Thanks for the quick reply! The version range is critical, so external developers know if they're on a supported path (and internal ones know which ruby features they can use).

It doesn't have to be `required_ruby_version` though, but please remove the entry if it won't be maintained (it's an [optional](https://guides.rubygems.org/specification-reference/) part of a gemspec).

**Comment by sambostock** [comment]
> The version range is critical

I agree with this. Honestly, I **personally** can't think of a reason why not to have the `required_ruby_version` reflect the versions these gems are tested against. To me, `required_ruby_version` is a commitment saying

> this gem works with the stated Ruby versions, and if it doesn't then a bug exists

As I mentioned in https://github.com/sorbet/sorbet/pull/4281#issuecomment-870939262, I think it is entirely Stripe's prerogative which Ruby versions are supported, but I think setting `required_ruby_version` to a version less than those that the gem is tested against is not an ideal choice as it can lead to the scenario discussed here: a breaking change was made for Ruby < 2.6 (the use of an endless change), but it was not caught by automated tests.

As for the alternative of removing `required_ruby_version` entirely, I don't think that would be a step in the right direction, as it conveys useful information to application developers and library authors, and removing it seems like a workaround to get away from a formally validated commitment to supported versions.

> the true test is whether it works, not whether the versions match.

While I agree with this in theory (it is indeed the _true_ test), I think there is value in having library authors explicitly exclude unsupported versions (which may incidentally work, but are not covered by tests and may break without warning at any time, as is the case for Ruby < 2.6 for Sorbet), because it saves everyone time to handle incompatibilities up front.
